android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PhysicalWebBroadcastService.java

```
@@ -0,0 +1,253 @@
+        Log.d(TAG, "SERVICE onStartCommand");
+        IntentFilter filter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
+        registerReceiver(mReceiver, filter);
+        mDisplayUrl = intent.getStringExtra(DISPLAY_URL_KEY);
+        Log.d(TAG, mDisplayUrl);
+
+        if (hasValidUrlLength(mDisplayUrl) && checkAndHandleAsciiUrl(mDisplayUrl)) {
+            // Set the url if we can
+            Log.d(TAG, "valid length");
+            mShareUrl = mDisplayUrl;
+            broadcastUrl();
+        } else {
+            Log.d(TAG, "needs shortening");
+            // Shorten the url if necessary
+            UrlShortenerClient.ShortenUrlCallback urlSetter =
+            new UrlShortenerClient.ShortenUrlCallback() {
+                @Override
+                public void onUrlShortened(String newUrl) {
+                    Log.d(TAG, "shortening success");
+                    mShareUrl = newUrl;
+                    broadcastUrl();
+                }
+                @Override
+                public void onError(String oldUrl) {
+                    Toast.makeText(getApplicationContext(), getString(R.string.shorten_error),
+                        Toast.LENGTH_LONG).show();
+                }
+            };
+            UrlShortenerClient.getInstance(this).shortenUrl(mDisplayUrl, urlSetter, TAG);
+        }
+        // Keep the service running
+        return START_STICKY;
+    }
+
+    private static boolean hasValidUrlLength(String url) {
+        int uriLength = UriBeacon.uriLength(url);
+        return 0 < uriLength && uriLength <= MAX_URI_LENGTH;
+    }
+
+    private boolean checkAndHandleAsciiUrl(String url) {
+        boolean isCompliant = false;
+        try {
+          URI uri = new URI(url);
+          String urlString = uri.toASCIIString();
+          isCompliant = url.equals(urlString);
+        } catch (URISyntaxException e) {
+            Toast.makeText(this, getString(R.string.nonascii_error), Toast.LENGTH_LONG).show();
```

Oh, this is for things that aren't URLs....depending on how airtight the search for a URL is, you could just throw a runtime error here.